File: Relink.Client/src/app/features/links/types/link.ts
export type Tag = {
    id: number;
    name: string;
};

export type Group = {
    id: number;
    name: string;
};

export type LinkMetadata = {
    id: number;
    shortenedLinkId: string;
    title: string | null;
    description: string | null;
    imageUrl: string | null;
    siteName: string | null;
    lastScrapedAt: string | null;
};

export type Link = {
    id: string;
    title: string;
    longUrl: string;
    createdAt: string;
    notes: string | null;
    startDate: string | null;
    expirationDate: string | null;
    passwordHash: string | null;
    maxVisits: number | null;
    visitCount: number;
    isLocked: boolean;

    tags?: Array<Tag>;
    metadata?: LinkMetadata | null;
    group?: Group | null;
    groupId?: number | null;
};

export type CreateLinkRequest = {
    longUrl: string;
    title: string;
    preferedShortCode?: string;
    notes?: string;
    startDate?: string;
    expirationDate?: string;
    password?: string;
    maxVisits?: number;
    tags?: string[];
    groupId?: number;
};

export type CreateLinkResponse = {
    shortCode: string;
    title: string;
};

export type UpdateLinkRequest = {
    longUrl: string;
    title: string;
    preferedShortCode?: string;
    notes?: string;
    startDate?: string | null;
    expirationDate?: string | null;
    password?: string | null;
    maxVisits?: number | null;
    tags?: string[];
    groupId?: number;
};

export type ScrapeMetadataResponse = Pick<LinkMetadata, 'title' | 'description' | 'imageUrl' | 'siteName'> & { lastScrapedAt: string };

export type VisitBucket = {
    start: string;
    end: string;
    count: number;
};

export type ReferrerCount = {
    referrer: string;
    count: number;
};

export type BrowserCount = {
    browser: string;
    count: number;
};

export type AnalyticsResponse = {
    visitCounts: Array<VisitBucket>;
    topReferrers: Array<ReferrerCount>;
    browserBreakdown: Array<BrowserCount>;
};

export type UnlockResponse = {
    longUrl: string;
};

File: Relink.ApiService/Data/Entities/Link.cs
namespace Relink.ApiService.Data.Entities;

public class Link
{
    public required string Id { get; set; }
    public string LongUrl { get; set; } = string.Empty;
    public string? Notes { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public string? FallbackUrl { get; set; }
    public DateTime? StartDate { get; set; }
    public DateTime? ExpirationDate { get; set; }
    public string? PasswordHash { get; set; }
    public int? MaxVisits { get; set; }
    public int VisitCount { get; set; } = 0;
    public bool IsLocked { get; set; } = false;

    public LinkMetadata? Metadata { get; set; }
    public List<LinkAnalytics> Analytics { get; set; } = [];
    public List<Tag> Tags { get; set; } = [];
    public Group? Group { get; set; }
    public int? GroupId { get; set; }
}